src/mcp/server/streamable_http_manager.py

```diff
@@ -233,12 +235,33 @@ async def run_server(*, task_status: TaskStatus[None] = anyio.TASK_STATUS_IGNORE
                     async with http_transport.connect() as streams:
                         read_stream, write_stream = streams
                         task_status.started()
-                        await self.app.run(
-                            read_stream,
-                            write_stream,
-                            self.app.create_initialization_options(),
-                            stateless=False,  # Stateful mode
-                        )
+                        try:
+                            await self.app.run(
+                                read_stream,
+                                write_stream,
+                                self.app.create_initialization_options(),
+                                stateless=False,  # Stateful mode
+                            )
+                        except Exception as e:
+                            logger.warning(
+                                f"Session {http_transport.mcp_session_id} crashed: {e}",
+                                exc_info=True,
+                            )
+                        finally:
+                            # Only remove from instances if not terminated
+                            if (
+                                http_transport.mcp_session_id
+                                and http_transport.mcp_session_id in self._server_instances
+                                and not (
+                                    hasattr(http_transport, "_terminated") and http_transport._terminated  # pyright: ignore
```

Comment:
To avoid having to override the linter, should we have `http_transport` expose a public property for checking termination state rather than directly access a private property?

```
  # streamable_http.py
  @property
  def is_terminated(self) -> bool:
      """Check if this transport has been explicitly terminated."""
      return getattr(self, '_terminated', False)
      
  # here
  if (
    http_transport.mcp_session_id
    and http_transport.mcp_session_id in self._server_instances
    and not http_transport.is_terminated
  ):
```